rcl/src/rcl/security.c

```diff
@@ -0,0 +1,350 @@
+// Copyright 2018 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "rcl/security.h"
+
+#include <stdbool.h>
+
+#include "rcl/error_handling.h"
+
+#include "rcutils/logging_macros.h"
+#include "rcutils/filesystem.h"
+#include "rcutils/format_string.h"
+#include "rcutils/get_env.h"
+#include "rcutils/strdup.h"
+
+#include "rmw/security.h"
+#include "rmw/security_options.h"
+
+#ifdef __clang__
+# pragma clang diagnostic push
+# pragma clang diagnostic ignored "-Wembedded-directive"
+#endif
+#include "tinydir/tinydir.h"
+#ifdef __clang__
+# pragma clang diagnostic pop
+#endif
+
+rcl_ret_t
+rcl_get_security_options_from_environment(
+  const char * name,
+  const char * namespace_,
+  const rcutils_allocator_t * allocator,
+  rmw_security_options_t * security_options)
+{
+  bool use_security = false;
+  rcl_ret_t ret = rcl_security_enabled(&use_security);
+  if (RCL_RET_OK != ret) {
+    return ret;
+  }
+
+  if (!use_security) {
+    security_options->enforce_security = RMW_SECURITY_ENFORCEMENT_PERMISSIVE;
+    return RMW_RET_OK;
+  }
+
+  RCUTILS_LOG_DEBUG_NAMED(
+    ROS_PACKAGE_NAME, "Using security: %s", use_security ? "true" : "false");
+
+  ret = rcl_get_enforcement_policy(&security_options->enforce_security);
+  if (RCL_RET_OK != ret) {
+    return ret;
+  }
+
+  // File discovery magic here
+  char * secure_root = rcl_get_secure_root(
+    name,
+    namespace_,
+    allocator);
+  if (secure_root) {
+    RCUTILS_LOG_INFO_NAMED(ROS_PACKAGE_NAME, "Found security directory: %s", secure_root);
+    security_options->security_root_path = secure_root;
+  } else {
+    if (RMW_SECURITY_ENFORCEMENT_ENFORCE == security_options->enforce_security) {
+      return RCL_RET_ERROR;
+    }
+  }
+  return RCL_RET_OK;
+}
+
+rcl_ret_t
+rcl_security_enabled(bool * use_security)
+{
+  const char * ros_security_enable = NULL;
+  const char * get_env_error_str = NULL;
+
+  RCL_CHECK_ARGUMENT_FOR_NULL(use_security, RCL_RET_INVALID_ARGUMENT);
+
+  get_env_error_str = rcutils_get_env(ROS_SECURITY_ENABLE_VAR_NAME, &ros_security_enable);
+  if (NULL != get_env_error_str) {
+    RCL_SET_ERROR_MSG_WITH_FORMAT_STRING(
+      "Error getting env var '" RCUTILS_STRINGIFY(ROS_SECURITY_ENABLE_VAR_NAME) "': %s\n",
+      get_env_error_str);
+    return RCL_RET_ERROR;
+  }
+
+  *use_security = (0 == strcmp(ros_security_enable, "true"));
```

Comment:
@ivanpauno meta: uh, this environment variable expects `true` as a boolean literal, while others above expect `1`. We should probably standardize that in a follow-up PR.

Reply:
It was already working like that. See that I first just moved code https://github.com/ros2/rcl/pull/515/commits/4aa706cf34cc69d2c86920c161c4aa964fb706a7.

I agree we should standarize. We should probably have a function in `rcutils` that accepts either "1" "true" "on" (not case sensitive).